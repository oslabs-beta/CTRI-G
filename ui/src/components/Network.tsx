// ---- imports go here ----
import {useState} from 'react'
import ContainerDisplay from './ContainerDisplay';
import type {
  ContainerInfo,
  NetworkInfo,
  setContainers,
  setNetworks,
} from '../interfaces/interfaces';
import { RemoveNetwork, HideContainers } from '../functions/functions';

// TO DO: typing will need to be more specific here once the exact contents of bridge and container are known
const Network = (props: {
  network: NetworkInfo;
  networkIndex: String;
  containers: ContainerInfo[] | [];
  setContainers: setContainers;
  setNetworks: setNetworks;
}) => {
  const [Loading, isLoading] = useState<Boolean>(false)
  // declare a variable, bridgeContainerDisplay, and assign it the value of an empty array
  const networkContainerDisplay: JSX.Element[] = [];
  // for each of the container objects passed down in the containerDisplay through props
  // TO DO: more specific typing on current container
  props.containers.forEach((currentContainer: ContainerInfo, i: number) => {
    // declare a variable, newContainer, and assign it the value of a container component passing down the currentContainer object as props
    if (props.network.Containers?.includes(currentContainer.Name)) {
      const newContainer = (
        <ContainerDisplay
          id={`${props.networkIndex}_container${i}`}
          key={`${props.networkIndex}_container${i}`}
          info={currentContainer}
          network={props.network.Name}
          setContainers={props.setContainers}
          setNetworks={props.setNetworks}
          isLoading={isLoading}
        />
      );
      // push the newContainer into the bridgeContainerDisplay
      networkContainerDisplay.push(newContainer);
    }
  });
  // TO DO: extract the bridge name from the bridge props
  // console.log('network: ', props.network);
  // console.log('networkContainerDisplay: ', networkContainerDisplay);
  if (networkContainerDisplay[0]) {
    // console.log(
    //   'networkContainerDisplay[0].props.info.Name: ',
    //   networkContainerDisplay[0].props.info.Name
    // );
  }
  // const bridgeName: string = bridgeContainerDisplay[0].props.info.name;
  const networkName = props.network.Name;
  // return
  return (
    // a div containing the bridge name and the array displaying each container
    <div id={`${props.networkIndex}`} className='network'>
      <div className='networkContainer'>
    { Loading?  (<p className= 'Loader'>Disconnecting...</p>):(
      <>
        <div className='networkLabel'>
          <strong>Name: </strong>
          {networkName}
        </div>
        <hr />
        <button
        className='innerButton'
        id={`${props.networkIndex}ShowHideNetworksButton`}
        onClick={() =>
          HideContainers(
            `${props.networkIndex}ContainersContainer`,
            `${props.networkIndex}ShowHideNetworksButton`,
            )
          }>
          Show Containers
        </button>
        <button
        className='deleteNetworkButton'
        onClick={() => RemoveNetwork(props.network.Name, props.setNetworks)}>
          x
        </button>
      <div
      id={`${props.networkIndex}ContainersContainer`}
      className='containersContainer'>
        {networkContainerDisplay}
        </div>
        </>
      )}
      </div>
      </div>
      
  );
};

// ---- exports go here ----
export default Network;
