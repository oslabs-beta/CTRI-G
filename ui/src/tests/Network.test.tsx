// add imports here
import Network from '../components/Network';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// mock the docker extension api client
jest.mock('@docker/extension-api-client', () => ({
  createDockerDesktopClient: jest.fn().mockReturnValue({
    docker: {
      cli: {
        exec: jest.fn(),
      },
    },
    desktopUI: {
      toast: {
        error: jest.fn(),
        success: jest.fn(),
        warning: jest.fn(),
      },
    },
  }),
}));

const ddClientMock = jest
  .requireMock('@docker/extension-api-client')
  .createDockerDesktopClient();

// set up the Network tests, clearing mocks and resetting modules before each
describe('Network component unit tests', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });
  // describe props for creating an empty bridge network
  describe('Network', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      jest.resetModules();
    });

    const emptyBridgeTestProps = {
      network: {
        Driver: 'bridge',
        Name: 'bridge',
        ID: '0',
        Containers: [],
      },
      networkIndex: 0,
      containers: [],
      allNetworks: [],
    };
    const emptyNoneTestProps = {
      network: {
        Driver: 'none',
        Name: 'none',
        ID: '0',
      },
      networkIndex: 0,
      containers: [],
      allNetworks: [],
    };
    const emptyHostTestProps = {
      network: {
        Driver: 'host',
        Name: 'host',
        ID: '0',
      },
      networkIndex: 0,
      containers: [],
      allNetworks: [],
    };
    const singleContainerTestProps = {
      network: {
        Driver: 'bridge',
        Name: 'Test Network',
        ID: '0',
        Containers: ['Port Navigator'],
      },
      networkIndex: 0,
      containers: [
        {
          Id: 'abc',
          Image: 'portNavigator',
          Name: 'Port Navigator',
          Networks: 'Test Network',
          Ports: null,
          State: 'Test',
        },
      ],
      allNetworks: [],
    };
    const multipleContainerTestProps = {
      network: {
        Driver: 'bridge',
        Name: 'Test Network',
        ID: '0',
        Containers: ['Port Navigator', 'Other Container'],
      },
      networkIndex: 0,
      containers: [
        {
          Id: 'abc',
          Image: 'portNavigator',
          Name: 'Port Navigator',
          Networks: 'Test Network',
          Ports: null,
          State: 'Test',
        },
        {
          Id: 'efg',
          Image: 'otherContainer',
          Name: 'Other Container',
          Networks: 'Test Network',
          Ports: null,
          State: 'Test',
        },
      ],
      allNetworks: [],
    };
    // const emptyUserCreatedBridgeProps = {
    //   network: {
    //     Driver: 'bridge',
    //     Name: 'Port Navigator',
    //     ID: '0',
    //     Containers: [],
    //   },
    //   containers: [],
    //   networkIndex: 0,
    //   allNetworks: [],
    // };

    // tests for the network name go here
    test('Should display the network name passed in props', () => {
      render(<Network {...singleContainerTestProps} />);
      const networkName = screen.getByText(/Test Network/);
      expect(networkName).toBeDefined();
    });

    // tests for the disconnect button go here
    test('Should display a disconnect button', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      expect(testDisconnectButton).toBeDefined();
    });

    test('Should prohibit deletion of the pre-generated bridge network', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      fireEvent.click(testDisconnectButton);
      expect(ddClientMock.desktopUI.toast.error).toHaveBeenCalledWith(
        "You can't delete the bridge network!"
      );
    });

    test('Should prohibit deletion of the pre-generated none network', () => {
      render(<Network {...emptyNoneTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      fireEvent.click(testDisconnectButton);
      expect(ddClientMock.desktopUI.toast.error).toHaveBeenCalledWith(
        "You can't delete the none network!"
      );
    });

    test('Should prohibit deletion of the pre-generated host network', () => {
      render(<Network {...emptyHostTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      fireEvent.click(testDisconnectButton);
      expect(ddClientMock.desktopUI.toast.error).toHaveBeenCalledWith(
        "You can't delete the host network!"
      );
    });

    test('Should prohibit deletion of user-created networks that hold a single container', () => {
      render(<Network {...singleContainerTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      fireEvent.click(testDisconnectButton);
      expect(ddClientMock.desktopUI.toast.error).toHaveBeenCalledWith(
        `You can't delete a Network that has Containers attached to it!`
      );
    });

    test('Should prohibit deletion of user-created networks that hold multiple containers', () => {
      render(<Network {...multipleContainerTestProps} />);
      const testDisconnectButton = screen.getByRole('button', {
        name: /x/,
      });
      fireEvent.click(testDisconnectButton);
      expect(ddClientMock.desktopUI.toast.error).toHaveBeenCalledWith(
        `You can't delete a Network that has Containers attached to it!`
      );
    });

    // test('Should allow deletion of user-created networks with no containers', async () => {
    //   render(<Network {...emptyUserCreatedBridgeProps} />);
    //   const testDisconnectButton = screen.getByRole('button', {
    //     name: /x/,
    //   });
    //   fireEvent.click(testDisconnectButton);
    //   await ddClientMock.desktopUI.toast.success;
    //   expect(ddClientMock.desktopUI.toast.success).toHaveBeenCalledWith(
    //     'Successfully deleted Network!'
    //   );
    // });

    // tests for the connected containers display go here
    test('Should display the correct number of containers when no containers a present', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const containerNumbers = screen.getByText(0);
      expect(containerNumbers).toBeDefined();
    });

    test('Should display the correct number of containers when one container is present', () => {
      render(<Network {...singleContainerTestProps} />);
      const containerNumbers = screen.getByText(1);
      expect(containerNumbers).toBeDefined();
    });

    test('Should display the connected containers', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const connectedContainersDivider = screen.getByText(/Connected/);
      expect(connectedContainersDivider).toBeDefined();
    });

    test('Should display the correct number of containers when no containers are connected', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const connectedContainersNum = screen.getByText(/0/);
      expect(connectedContainersNum).toBeDefined();
    });

    test('Should display the correct number of containers when one container is connected', () => {
      render(<Network {...singleContainerTestProps} />);
      const connectedContainersNum = screen.getByText(/1/);
      expect(connectedContainersNum).toBeDefined();
    });

    test('Should display the correct number of containers when multiple containers are connected', () => {
      render(<Network {...multipleContainerTestProps} />);
      const connectedContainersNum = screen.getByText(/2/);
      expect(connectedContainersNum).toBeDefined();
    });

    // tests for the show/hide containers button go here
    test('Should not display a show containers button for networks without containers', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const showContainersButton = screen.queryByRole('button', {
        name: /Show Containers/,
      });
      expect(showContainersButton).toBeNull();
    });

    test('Should not display a hide containers button for networks without containters', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const hideContainersButton = screen.queryByRole('button', {
        name: /Hide Containers/,
      });
      expect(hideContainersButton).toBeNull();
    });

    test('Should display a show containers button when the network is first created if the network has containers', () => {
      render(<Network {...singleContainerTestProps} />);
      const showContainersButton = screen.queryByRole('button', {
        name: /Show Containers/,
      });
      expect(showContainersButton).toBeDefined();
    });

    // test('Should display a hide containers button after the show containers button is clicked', () => {
    //   render(<Network {...singleContainerTestProps} />);
    //   const showContainersButton = screen.getByRole('button', {
    //     name: /Show Containers/,
    //   });
    //   fireEvent.click(showContainersButton);
    //   const hideContainersButton = screen.getByText(/Hide Containers/);
    //   expect(hideContainersButton).not.toBeNull();
    // });

    test('Should not have visible containers when the component is initially rendered', () => {
      render(<Network {...singleContainerTestProps} />);
      const containerName = screen.queryByText(/Port Navigator/);
      expect(containerName).not.toBeVisible();
    });

    test('Should make containers visible when the show containers button is clicked', () => {
      render(<Network {...singleContainerTestProps} />);
      const showContainersButton = screen.getByRole('button', {
        name: /Show Containers/,
      });
      const containerName = screen.getByText(/Port Navigator/);
      fireEvent.click(showContainersButton);
      expect(containerName).toBeVisible();
    });

    test('Should make containers invisible when the hide containers button is clicked', () => {
      render(<Network {...singleContainerTestProps} />);
      const showContainersButton = screen.getByRole('button', {
        name: /Show Containers/,
      });
      const containerName = screen.getByText(/Port Navigator/);
      fireEvent.click(showContainersButton);
      fireEvent.click(showContainersButton);
      expect(containerName).not.toBeVisible();
    });

    // tests for the add container button go here
    test('Should display an add container button', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const addContainerButton = screen.getByRole('button', {
        name: /Add Container/,
      });
      expect(addContainerButton).toBeDefined();
    });

    test('Should display the add container form when the add container button is clicked', () => {
      render(<Network {...emptyBridgeTestProps} />);
      const addContainerButton = screen.getByRole('button', {
        name: /Add Container/,
      });
      fireEvent.click(addContainerButton);
      const connectNetworkPhrase = screen.getByText(/container to the/);
      expect(connectNetworkPhrase).toBeDefined();
    });
  });
});
