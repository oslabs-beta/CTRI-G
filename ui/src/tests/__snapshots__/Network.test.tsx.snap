// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network component unit tests Network renders Network 1`] = `
<div
  className="userNetwork"
>
  <div
    className="networkContainer"
  >
    <div
      className="networkLabel"
    >
      <strong>
        Network: 
      </strong>
      bridge
    </div>
    <hr
      className="hrNet"
    />
    <div
      className="networkInfo"
    >
      <div
        className="Driver"
      >
        <strong>
          Driver: 
        </strong>
        bridge
      </div>
      <div
        className="Gateway"
      >
        <strong>
          Gateway: 
        </strong>
        null
      </div>
      <div
        className="Subnet"
      >
        <strong>
          Subnet: 
        </strong>
        null
      </div>
    </div>
    <hr />
    <div
      className="containerNetworkFeatures"
    >
      <div
        className="connectedContainerContainer"
      >
        <p
          className="connectedText"
        >
          Connected
          <br />
          Containers
        </p>
        <p
          className="divider"
        >
          |
        </p>
        <p
          className="connectedCount"
        >
          0
        </p>
      </div>
      <div />
      <button
        className="innerButton"
        id="0ShowHideNetworksButton"
        onClick={[Function]}
      >
        Add Container
      </button>
    </div>
    <button
      className="deleteNetworkButton"
      onClick={[Function]}
    >
      x
    </button>
  </div>
  <div
    className="containersContainer"
    id="0ContainersContainer"
    style={
      {
        "display": "none",
      }
    }
  />
</div>
`;
